WIP: Further improving fast date parsing

omus opened the pull request:
Iterating on #18000 and #19545. Speed up is currently 140x.

before:

```julia
julia> @benchmark DateTime("2010-10-10T10:10:10.10")
BenchmarkTools.Trial: 
  memory estimate:  3.11 kb
  allocs estimate:  85
  --------------
  minimum time:     16.768 μs (0.00% GC)
  median time:      17.943 μs (0.00% GC)
  mean time:        20.198 μs (1.78% GC)
  maximum time:     3.734 ms (96.53% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

Shashi's update #19545

```julia
julia> @benchmark DateTime("2010-10-10T10:10:10.10")
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     138.344 ns (0.00% GC)
  median time:      140.283 ns (0.00% GC)
  mean time:        157.004 ns (0.00% GC)
  maximum time:     1.683 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     861
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

This PR:
```julia
julia> @benchmark DateTime("2010-10-10T10:10:10.10")
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     118.559 ns (0.00% GC)
  median time:      125.402 ns (0.00% GC)
  mean time:        138.223 ns (0.00% GC)
  maximum time:     2.867 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     907
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

fixes #15888, #13644
supersedes #18000, #19545

omus commented:
I'm also experimenting with some ideas to apply this performance gain to all date formats. Feel free to merge in the mean time.

simonbyrne commented:
@shashi?

omus commented:
Modified direction of the PR to work off of #19545. Mostly I am cleaning up the code without losing the speedup. I will probably also address #16928 here and may also include some custom code for specific formats.

shashi commented on base/dates/io-util.jl:
```diff
@@ -129,12 +118,11 @@ end
-    while (maxchars <= 0 || j <= maxchars) && i <= len
+    max_pos = maxchars <= 0 ? len : min(chr2ind(str, ind2chr(str,i) + maxchars - 1), len)
+    while i <= max_pos
         c, ii = next(str, i)
```
you can actually also put `@inbounds` on this line and pretty much everywhere `next` is called to squeeze out a few more nanoseconds!

shashi reviewed (commented):
@omus were you able to accomodate the changes required for TimeZones? I see that things still deal with NTupl{7,Int}s...

I have a new commit, should try and merge it here.

shashi commented:
I think this can be done by something like adding a parameter to `DatePart{char, T}` where T is the output type, in this case the type of the result in `_tryparse` can still be determined... You might have other ideas.

omus commented:
@shashi I haven't yet dealt with parsing support for TimeZones. I have some ideas including what you suggested but I'll need to do more experimentation to find out what works best.

omus commented:
Added fundamentals of adding custom slots to `TimeType` parsing (supports parsing of `ZonedDateTime` type). Changes don't impact performance of `Date` or `DateTime`

omus committed:
Refactoring some tests

omus committed:
Dispatch on custom parser for improved performance

```julia
julia> s = "2010-10-11T12:13:14.15"
"2010-10-11T12:13:14.15"

julia> df_default = Base.Dates.DateFormat("yyyy-mm-ddTHH:MM:SS.s")
dateformat"yyyy-mm-ddTHH:MM:SS.s"

julia> df_custom = Base.Dates.DateFormat("yyyy-mm-dd\\THH:MM:SS.s")
dateformat"yyyy-mm-ddTHH:MM:SS.s"

julia> @benchmark parse(DateTime,$s,$df_default)
BenchmarkTools.Trial:
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     127.267 ns (0.00% GC)
  median time:      135.160 ns (0.00% GC)
  mean time:        147.785 ns (0.00% GC)
  maximum time:     2.450 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     886
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark parse(DateTime,$s,$df_custom)
BenchmarkTools.Trial:
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     71.947 ns (0.00% GC)
  median time:      74.662 ns (0.00% GC)
  mean time:        82.548 ns (0.00% GC)
  maximum time:     2.529 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     976
  time tolerance:   5.00%
  memory tolerance: 1.00%

```

shashi commented:
Nice! :)

Is ZonedDateTime parsing itself fast?

omus commented:
ZonedDateTime parsing is about 10x faster. There are probably some additional things I can do outside of base to improve the performance more.

omus commented:
I've got one more change I'd like to do which will be done on Monday.

omus committed:
Minor refactor to DateFormat constructor

omus committed:
Update testcases

omus commented:
I've added support for custom parsers which now makes DateTime parsing using ISODateTimeFormat 230x faster.

```julia
julia> @benchmark DateTime("2010-10-10T10:10:10.10")
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     70.023 ns (0.00% GC)
  median time:      77.467 ns (0.00% GC)
  mean time:        109.659 ns (0.00% GC)
  maximum time:     9.145 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     976
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

omus commented:
@shashi feel free to merge in my changes again. Code should be ready for review.

tkelman commented:
should we change the target branch of this pr?

omus commented:
Sure. Do I need to remove the "fixes" / "supersedes" notes before merging?

tkelman commented:
good question, don't think so - i believe github only auto closes things that merge to master

tkelman commented on test/dates/io.jl:
```diff
@@ -66,11 +66,15 @@ b2 = "96/Feb/1"
 @test_throws ArgumentError Dates.DateTime(b3,f)
-let err = try DateTime("2012-02-20T09:09:3.43i9") catch err; err end
+try
+    Dates.tryparse_internal(DateTime, "2012-02-20T09:09:3.43i9", Dates.ISODateTimeFormat, true)
```
add a `@test false` here if the exception should happen all the time, or if it doesn't throw then this would pass silently

omus committed:
Avoid passing test silently if no exception thrown

shashi commented:
:+1:
